### PR TITLE
Add Funnel Discovery class

### PR DESCRIPTION
**Pull request by vard-obj**
TODO: add `funnel-discovery` notebook in the docs

### CODE REVIEW

File: modelhub/modelhub/modelhub.py [hunk 1]
```diff
@@ -272,3 +273,13 @@ def get_logistic_regression(self, *args, **kwargs) -> LogisticRegression:
         """
 
         return LogisticRegression(*args, **kwargs)
+
+    def get_funnel_discovery(self) -> FunnelDiscovery:
+        """
+        Return an instance of the :py:class:`FunnelDiscovery` from the model hub.
+
+        All parameters passed to this function are passed to the constructor of the FunnelDiscovery
+        model.
```
Comment (KathiaBarahona) [inline modelhub/modelhub/modelhub.py:282]: Not really, function is not receiving any param
```suggestion
```

File: modelhub/modelhub/models/funnel_discovery.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+        if not all(col in conversion_events_df.data_columns
+                   for col in conversion_events_df_columns):
+            raise ValueError(f'The {", ".join(conversion_events_df_columns)} columns should be in the main dataframe')
+
+        _steps_df = steps_df.copy()
+        _steps_df['_index'] = steps_df.groupby().window().location_stack_step_1.window_row_number()
+
+        # if in get_navigation_paths groupby used, we have to add it in the dataframe index
+        index_columns = ['_index']
+        if self.groupby is not None and self.groupby is not not_set:
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:185]: not really, for example:

```
df_steps = FunnelDiscovery().get_navigation_paths(..., by='user_id')

# new FunnelDiscovery instance with empty groupby
filtered_df_steps = FunnelDiscovery().filter_navigation_paths_to_conversion(df_steps, ...)
```

File: modelhub/modelhub/models/funnel_discovery.py [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+        result = result[result[offset_lc.name] < steps]
+
+        # re-order rows
+        result = result.sort_values(by=result.index_columns + [offset_lc.name])
+        result = result.drop(columns=[offset_lc.name])
+        return result
+
+    def filter_navigation_paths_to_conversion(self, steps_df, conversion_events_df,
+                                              conversion_location_column='feature_nice_name'
+                                              ) -> bach.DataFrame:
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:155]:
```suggestion
    def filter_navigation_paths_to_conversion(
        self, 
        steps_df: bach.DataFrame,
        conversion_events_df: bach.DataFrame,
        conversion_location_column: str = 'feature_nice_name'
    ) -> bach.DataFrame:
```

File: modelhub/modelhub/models/funnel_discovery.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+
+        :param steps_df: dataframe which one gets from `get_navigation_paths` method output.
+            This the dataframe that we're going to filter.
+        :param conversion_events_df: dataframe where for each `conversion_location_column`
+            value we have info if there was a conversion event. We will merge this dataframe
+            with 'melted' steps_df in order to have `is_conversion_event` column for steps_df.
+        :param conversion_location_column: need this column in the dataframes merge in order
+            to get converted steps df.
+
+        :returns: bach DataFrame, filtered each row of steps_df to the
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:172]: :P probably this is missing some words

File: modelhub/modelhub/models/funnel_discovery.py [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+        _location_stack = location_stack or data['location_stack']
+        _location_stack = _location_stack.copy_override_type(SeriesLocationStack)
+        partition = None
+        sort_nice_names_by = []
+
+        from modelhub.aggregate import Aggregate
+        if self.groupby is not None and self.groupby is not not_set:
+            partition = Aggregate._check_groupby(
+                data=data, groupby=self.groupby, not_allowed_in_groupby='location_stack',
+            )
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:82]: mmm let's avoid calling private methods, probably `_check_groupby` should be a public method in `modelhub.utils`

File: modelhub/modelhub/models/funnel_discovery.py [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+            then steps are based on the order of events based on the entire dataset.
+        :param location_stack: the location stack
+
+            - can be any slice of a :py:class:`modelhub.SeriesLocationStack` type column
+            - if None - the whole location stack is taken.
+
+        :returns: bach DataFrame containing a new series for each step containing the nice name
+            of the location.
+        """
+        self.groupby = by
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:70]: I don't think you actually need `self.groupby`

File: modelhub/modelhub/models/funnel_discovery.py [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+
+        from modelhub.series.series_objectiv import SeriesLocationStack
+        _location_stack = location_stack or data['location_stack']
+        _location_stack = _location_stack.copy_override_type(SeriesLocationStack)
+        partition = None
+        sort_nice_names_by = []
+
+        from modelhub.aggregate import Aggregate
+        if self.groupby is not None and self.groupby is not not_set:
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:79]:
```suggestion
        if by is not None and by is not not_set:
```

File: modelhub/modelhub/models/funnel_discovery.py [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+        _steps_df = _steps_df.reset_index().set_index(index_columns)
+
+        def tag_step(step_series: bach.SeriesString) -> bach.Series:
+            # tags the step number, is it 1st step, 2nd step, ...
+            step_df = step_series.to_frame()
+            # add the step name as index, this way we can identify it later
+            step_df['step_tag'] = int(step_series.name.split('_')[-1])
+            step_df = step_df.set_index(keys='step_tag', append=True)
+            step_df = step_df.materialize(node_name='tagged_step')
+            return step_df[step_series.name]
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:197]: This should be a `staticmethod` in `FunnelDiscovery` class

File: modelhub/modelhub/models/funnel_discovery.py [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+        :returns: bach DataFrame, filtered each row of steps_df to the
+        """
+
+        conversion_events_df_columns = [conversion_location_column, 'is_conversion_event']
+        if not all(col in conversion_events_df.data_columns
+                   for col in conversion_events_df_columns):
+            raise ValueError(f'The {", ".join(conversion_events_df_columns)} columns should be in the main dataframe')
+
+        _steps_df = steps_df.copy()
+        _steps_df['_index'] = steps_df.groupby().window().location_stack_step_1.window_row_number()
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:181]: I will be carefull with this line, what if the location stack event has actually a different name?
```suggestion
        _steps_df['_index'] = steps_df.groupby().window().location_stack_step_1.window_row_number()
```

File: modelhub/modelhub/models/funnel_discovery.py [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+
+        # if in get_navigation_paths groupby used, we have to add it in the dataframe index
+        index_columns = ['_index']
+        if self.groupby is not None and self.groupby is not not_set:
+            index_columns.append(self.groupby)
+
+        _steps_df = _steps_df.reset_index().set_index(index_columns)
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:188]: This is going to break if `_index` is not part of the `_steps_df`. I think you can know how was  `steps_df` grouped based on the series in the index

File: modelhub/modelhub/models/funnel_discovery.py [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+
+        # limit rows where initial step is between the requested steps
+        result = result[result[offset_lc.name] < steps]
+
+        # re-order rows
+        result = result.sort_values(by=result.index_columns + [offset_lc.name])
+        result = result.drop(columns=[offset_lc.name])
+        return result
+
+    def filter_navigation_paths_to_conversion(self, steps_df, conversion_events_df,
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:153]: **suggestion** Maybe this function can be separated into other methods, as it's a bit complex and hard to handle

File: modelhub/modelhub/models/funnel_discovery.py [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+                                              conversion_location_column='feature_nice_name'
+                                              ) -> bach.DataFrame:
+        """
+        Filter steps dataframe rows to first conversion location.
+        For each row of steps_df dataframe set to None the step values
+         after encountering the first conversion location step.
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:159]: Maybe a bit more explanation about the steps will be good, as its not clear what the function actually does

File: modelhub/modelhub/models/funnel_discovery.py [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+        # re-order rows
+        result = result.sort_values(by=result.index_columns + [offset_lc.name])
+        result = result.drop(columns=[offset_lc.name])
+        return result
+
+    def filter_navigation_paths_to_conversion(self, steps_df, conversion_events_df,
+                                              conversion_location_column='feature_nice_name'
+                                              ) -> bach.DataFrame:
+        """
+        Filter steps dataframe rows to first conversion location.
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:157]: Maybe?
```suggestion
        Identify the first conversion step per each navigation path.
```

File: modelhub/modelhub/models/funnel_discovery.py [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+        For each row of steps_df dataframe set to None the step values
+         after encountering the first conversion location step.
+
+        Corner case: in case the first location stack is conversion one we ignore it
+        and proceed to the next step.
+
+        :param steps_df: dataframe which one gets from `get_navigation_paths` method output.
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:164]: :P maybe some reference to the method, for docs

File: modelhub/modelhub/models/funnel_discovery.py [hunk 15]
```diff
@@ -0,0 +1,329 @@
...
+        if self.groupby is not None and self.groupby is not not_set:
+            result_df = result_df.set_index(self.groupby)
+
+        return result_df
+
+    def plot_sankey_diagram(self, steps_counter_df: bach.DataFrame, n_top_examples=15, max_n_top_examples=50):
+        """
+        Plot Sankey Diagram.
+
+        :param steps_counter_df: constructing pandas links dataframe out of steps_counter_df.
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:247]: not sure if I understand this :thinking:

**Review by KathiaBarahona — CHANGES_REQUESTED** [review]
left some comments, I haven't take a look to the notebook. It's better to have a separate PR just for it. Because maybe some of this methods can be private and just provide the params to the sankey diagram :thinking: . But have to see more how they are been used.

File: modelhub/modelhub/models/funnel_discovery.py (hunk 5 above)
Comment (vard-obj) [reply modelhub/modelhub/models/funnel_discovery.py]: Good idea, will create a separate PR.

File: modelhub/modelhub/models/funnel_discovery.py [hunk 16]
```diff
@@ -0,0 +1,414 @@
...
+    from modelhub.series import SeriesLocationStack
+
+
+GroupByType = Union[List[Union[str, Series]], str, Series, NotSet]
+
+
+class FunnelDiscovery:
+    """
+
+    """
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:24]: missing docstring:)

File: modelhub/modelhub/models/funnel_discovery.py [hunk 17]
```diff
@@ -0,0 +1,414 @@
...
+    def __init__(self):
+        pass
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:26]:
```suggestion
```

File: modelhub/modelhub/models/funnel_discovery.py [hunk 18]
```diff
@@ -0,0 +1,414 @@
...
+        result = result.sort_values(by=result.index_columns + [offset_lc.name])
+        result = result.drop(columns=[offset_lc.name])
+        return result
+
+    @staticmethod
+    def tag_step(step_series: bach.Series) -> bach.Series:
+        """
+        Tag the step number, is it 1st step, 2nd step, ...
+
+        @param step_series: series of ith step:
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:161]:
```suggestion
        :param step_series: series of ith step:
```

File: modelhub/modelhub/models/funnel_discovery.py [hunk 19]
```diff
@@ -0,0 +1,414 @@
...
+        if n_examples is not None:
+            result = result[result[offset_lc.name] < n_examples]
+
+        # re-order rows
+        result = result.sort_values(by=result.index_columns + [offset_lc.name])
+        result = result.drop(columns=[offset_lc.name])
+        return result
+
+    @staticmethod
+    def tag_step(step_series: bach.Series) -> bach.Series:
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:157]: let's make this private
```suggestion
    def _tag_step(step_series: bach.Series) -> bach.Series:
```

File: modelhub/modelhub/models/funnel_discovery.py [hunk 20]
```diff
@@ -0,0 +1,414 @@
...
+        """
+
+        step_df = step_series.to_frame()
+        # add the step name as index
+        step_df['step_tag'] = int(step_series.name.split('_')[-1])
+        step_df = step_df.set_index(keys='step_tag', append=True)
+        step_df = step_df.materialize(node_name='tagged_step')
+        return step_df[step_series.name]
+
+    def melt_steps(self, steps_df: bach.DataFrame) -> bach.Series:
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:185]: same here 
```suggestion
    def _melt_steps(self, steps_df: bach.DataFrame) -> bach.Series:
```

File: modelhub/modelhub/models/funnel_discovery.py [hunk 21]
```diff
@@ -0,0 +1,414 @@
...
+        and proceed to the next step.
+
+        :param steps_df: dataframe which one gets from `FunnelDiscovery.get_navigation_paths` method,
+            the dataframe that we're going to filter.
+
+        :returns: bach DataFrame, filtered each row of steps_df to the conversion location.
+        """
+
+        conv_step_num_column = 'first_conversion_step_number'
+        result_df = steps_df[steps_df[conv_step_num_column] != 1]
```
Comment (KathiaBarahona) [inline modelhub/modelhub/models/funnel_discovery.py:303]: what if `conv_step_num_column` does not exist in `steps_df`?

**Review by KathiaBarahona — COMMENTED** [review]
Left a couple of comments. I think my major concern is that most of this functions expect a dataframe with certain structure but its not quite clear which are the steps to follow for generating such df.

Also, I didn't went through deeply to `plot_sankey_diagram`, is it necessary to use pandas?

**Pull request closed without merging**